src/kernel/heap.zig

```diff
@@ -0,0 +1,388 @@
+        }
+    }
+};
+
+test "init errors on non-power-of-two" {
+    const start = 10;
+    const allocator = std.heap.page_allocator;
+    // Non-power-of-size heap size
+    testing.expectError(Error.NonPowerOfTwo, Heap.init(start, 100, 1024, allocator));
+    // Non-power-of-two min block size
+    testing.expectError(Error.NonPowerOfTwo, Heap.init(start, 1024, 100, allocator));
+    // Non-power-of-two heap size and min block size
+    testing.expectError(Error.NonPowerOfTwo, Heap.init(start, 100, 100, allocator));
+    // Power-of-two heap size and min block size
+    var heap = try Heap.init(start, 1024, 1024, allocator);
+}
+
+test "free detects unallocated addresses" {
+    var heap = try Heap.init(10, 1024, 16, std.heap.page_allocator);
+    // Before start of heap
+    testing.expectError(Error.NotAllocated, heap.free(0, heap.block_min_size));
+    // At start of heap
+    testing.expectError(Error.NotAllocated, heap.free(heap.start, heap.block_min_size));
+    // Within the heap
+    testing.expectError(Error.NotAllocated, heap.free(21, heap.block_min_size));
+    // End of heap
+    testing.expectError(Error.NotAllocated, heap.free(heap.start + heap.size, heap.block_min_size));
+    // Beyond heap
+    testing.expectError(Error.NotAllocated, heap.free(heap.start + heap.size + 1, heap.block_min_size));
+}
+
+test "whole heap can be allocated and freed" {
+    var heap = try Heap.init(0, 1024, 16, std.heap.page_allocator);
+    var occupied: usize = 0;
+    var rand = std.rand.DefaultPrng.init(123).random;
+
+    // Allocate entire heap
+    while (occupied < heap.size) {
+        // This allocation should succeed
+        const result = heap.alloc(heap.block_min_size, null) orelse unreachable;
+        testing.expectEqual(occupied, result);
+        occupied += heap.block_min_size;
+    }
+    // No more allocations should be possible
+    testing.expectEqual(heap.alloc(1, null), null);
+
+    // Try freeing all allocations
+    while (occupied > 0) : (occupied -= heap.block_min_size) {
+        const addr = occupied - heap.block_min_size;
+        heap.free(addr, heap.block_min_size) catch unreachable;
+        // Make sure it can be reallocated
+        const result = heap.alloc(heap.block_min_size, null) orelse unreachable;
+        testing.expectEqual(addr, result);
+        // Re-free it
+        heap.free(addr, heap.block_min_size) catch unreachable;
+    }
+
+    // Trying to free any previously allocated address should now fail
+    var addr: usize = 0;
+    while (addr < heap.size) : (addr += heap.block_min_size) {
+        testing.expectError(Error.NotAllocated, heap.free(addr, heap.block_min_size));
+    }
+
+    // The bitmap should now be clear, otherwise free didn't clean up properly
+    try testBitmapClear(&heap);
+}
+
+test "Allocator" {
+    var buff = [_]u8{0} ** (4 * 1024 * 1024);
+    var heap = try Heap.init(@ptrToInt(&buff), buff.len, 16, std.heap.page_allocator);
```

Comment:
Could try other block sizes and could run some bench marks to see what speed difference you get

Reply:
Sure thing, that would be interesting. I guess that's something we could do on the side outside of this PR.